test/boost/rest_authenticator_test.cc

```diff
@@ -0,0 +1,464 @@
+            };
+
+            BOOST_REQUIRE_EXCEPTION(a.authenticate(creds).get(), exceptions::authentication_exception,
+                                    seastar::testing::exception_predicate::message_contains("Bad password"));
+
+            // Check state in DB
+            auto &qp = env.local_qp();
+            BOOST_REQUIRE(can_login(qp, "foo.bar").get() == false);
+
+            // Check state through role_manager should be align with DB state
+            auto &rm = env.local_auth_service().underlying_role_manager();
+            BOOST_REQUIRE_EXCEPTION(rm.can_login("foo.bar").get(), exceptions::invalid_request_exception,
+                                    seastar::testing::exception_predicate::message_contains(
+                                            "Role foo.bar doesn't exist."));
+        });
+}
+
+
+SEASTAR_TEST_CASE(user_has_roles) {
+        return with_dummy_authentication_server([](cql_test_env &env) {
+            auto &a = env.local_auth_service().underlying_authenticator();
+
+            auto creds = auth::authenticator::credentials_map{
+                    {auth::authenticator::USERNAME_KEY, sstring("alice")},
+                    {auth::authenticator::PASSWORD_KEY, sstring("password")}
+            };
+
+            a.authenticate(creds).discard_result().get();
+
+            auth::role_set roles;
+            roles.insert(sstring("scylla-rw"));
+            roles.insert(sstring("other"));
+
+            // Check state in DB
+            auto &qp = env.local_qp();
+            BOOST_REQUIRE_EQUAL(get_role_set(qp, "alice").get(), roles);
+
+            // Check state through role_manager
+            roles.insert(sstring("alice")); // query_granted also return current role name in the role set
+            auto &rm = env.local_auth_service().underlying_role_manager();
+            BOOST_REQUIRE_EQUAL(rm.query_granted("alice", auth::recursive_role_query::no).get(), roles);
+        });
+}
+
+SEASTAR_TEST_CASE(user_expired_is_well_recreated) {
+        return with_dummy_authentication_server([](cql_test_env &env) {
+            auto &a = env.local_auth_service().underlying_authenticator();
+
+            auto creds = auth::authenticator::credentials_map{
+                    {auth::authenticator::USERNAME_KEY, sstring("alice")},
+                    {auth::authenticator::PASSWORD_KEY, sstring("password")}
+            };
+
+            a.authenticate(creds).discard_result().get();
+
+            auto &qp = env.local_qp();
+            BOOST_REQUIRE_EQUAL(require_record_valid(qp, "alice").get(), "alice");
+            BOOST_REQUIRE_EXCEPTION(require_record_valid(qp, "foo.bar").get(), exceptions::invalid_request_exception,
+                                    seastar::testing::exception_predicate::message_contains(
+                                            "Role foo.bar doesn't exist."));
+
+            // Simulate TTL deleting the record
```

Comment:
I'm wondering if we should add one more test to validate that TTL is working as expected.